### ISSUE TITLE
Closing viewer programmatically

### ISSUE THREAD

**Issue by FedeClaudi** [napari/napari#1057]
## 🚀 Feature
I can't find a way to close the window programmatically, but it'd be a useful feature to have.

## Motivation
I'm writing a [small python package](https://github.com/BrancoLab/imgreg2D) and in it I need to create and close napari viewers as needed. To make things easier for the user I'd like to close the viewer programmatically when an operation is completed, rather than requiring the user to press on the 'x' button. This makes the flow of the application smoother and faster and reduces the risk of user errors.

## Pitch
add a `.close()` method that closes the viewer as needed. 

## Alternatives
Not doing and leaving just the 'x' button.

## Additional context
I've tried using this:
```
     viewer.window._qt_window.close()
```
but this doesn't work, it leaves the viewer open and unresponsive to button presses.

Minimal code to reproduce:
```
import napari

with napari.gui_qt():
    v1 = napari.Viewer()

    v1.window._qt_window.close()

with napari.gui_qt():
    v2 = napari.Viewer()
```

`v2` is never displayed because `v1` doesn't close properly. 
This was with napari v 0.2.12 on macOS.

**Comment by tlambert03** [comment]
Hi @FedeClaudi,
We revamped the shutdown process and added a `viewer.close()` method in #1014.   Try the current master branch at let us know if it works for you.

**Comment by tlambert03** [comment]
I'll also add that part of what you're seeing here might have to do with using multiple context managers like that in a script.  I don't have a full explanation... but there's obviously some race conditions that happen when spinning up and shutting down a Qt event loop.  So you might run into weirdness when using two `napari.gui_qt()` calls like that in sequence in a script.  

Even on the master branch, creating a viewer and immediately closing it like this has issues:
```python
import napari 

with napari.gui_qt(): 
    v1 = napari.Viewer(title='v1') 
    v1.close() 
```

but this works fine and shows viewer 2:
```python
import napari 

with napari.gui_qt(): 
    v1 = napari.Viewer(title='1') 
    v1.close()
    v2. = napari.Viewer(title='v2')
```

part of this is probably stuff we can fix (though it's proven very hard to debug), but the other aspect here is just using Qt a little differently, and not over-taxing the spin-up/shut-down proces.

**Comment by FedeClaudi** [comment]
Hi, 

Thanks for getting back to me so quickly. 
I've installed from the master branch and tried it.
There is a close method but it doesn't seem to work. 

This is the result:
<img width="1287" alt="image" src="https://user-images.githubusercontent.com/17436313/76764419-b6500000-678c-11ea-8edc-da177c725785.png">

As far as I can tell, `viewer.close()` just calls `self.window.close()` so it's the same as what I was trying earlier.

**Comment by tlambert03** [comment]
sorry, see my previous comment... `viewer.close()` _does_ call `viewer.window.close()`. but #1014 significantly changed `viewer.window.close()` as well.  So the master branch is _not_ the same as v0.2.12 .... however, the bigger issue here is the immediate opening and closing of a viewer in a context manager.

**Comment by FedeClaudi** [comment]
Hi, 

This solution:
```
import napari 

with napari.gui_qt(): 
    v1 = napari.Viewer(title='1') 
    v1.close()
    v2. = napari.Viewer(title='v2')
```

works, but I can't use it in my application. 
The two viewers are created by different functions at different times, so I need to call `with napari.gui_qt():` twice.



---
p.s. I've installed from the master branch with:
```
pip install git+https://github.com/napari/napari
```
is there something else I should try?

**Comment by tlambert03** [comment]
`napari.gui_qt()` is only a convenience for spinning up a quick Qt event loop.  If you need a more persistent Qt application (with which to run multiple different functions), then you should be starting it elsewhere. (for instance, using `ipython --gui=qt`, using the `%gui qt` IPython magic, or calling `QtWidgets.QApplication()` directly in your program before using napari).  the most appropriate choice will depend on the specifics of your application.
If you can direct me to the "different functions at different times" in your package, I might be able to provide a suggestion

**Comment by FedeClaudi** [comment]
Awesome thanks!

So the users need to first click on a set of points on the reference image used for image registration.
This is done using `get_fixed_points` in `imgreg2D.points.py` (https://github.com/BrancoLab/imgreg2D/blob/master/imgreg2D/points.py)

After that the user needs to click on the same points on the image being aligned, this is done with `get_registering_points`, in the same script. 

The reason why I want them to be separate, is that the user might have pre-defined `fixed_points` that he/she wants to use for the registration, so I want to give the option to skip the `get_fixed_points` step. Alternatively, one might just want to get the `fixed_points` without  doing the registration, so again it's useful to have the two steps separate.

I'd rather not use ipython/jupyter if possible, but I'm not very familiar with Qt application workflow. Any help would be appreciated, thank you!

**Comment by FedeClaudi** [comment]
Hi,

So, 
```
import napari

with napari.gui_qt():
    v1 = napari.Viewer()

    v1.window._qt_window.close()

with napari.gui_qt():
    v2 = napari.Viewer()
```
Doesn't work.

But I decided to try `viewer.close()` in my code anyways, and there it works!
Not sure if it's becaues of the delay between the two calls to `with napari.gui_qt(): ` or something else, but at least in my application `viewr.close()` seems to be working perfectly. 

Thank you very much for the help!

**Comment by tlambert03** [comment]
glad to hear it.  
still happy to show you an alternative to using `gui_qt()` throughout your code.  
It's a bit hard from your readme to figure out exactly how you're calling all your functions (and the GIF isn't high enough resolution).  How do you actually launch your program?  and does it involve multiple "executions", or is the full "run" contained in a single session?
It looks like you're using the VScode launch button on `test.py` ... but test.py doesn't have a `if __main__` statement.  what do you have in your launch.json?

**Comment by tlambert03** [comment]
a general pattern you might consider is to try to design your program such that you only need a single `with napari.gui_qt` call... so, for instance, your `test.py` might look like this instead:
```python
from imgreg2D.register import register
from imgreg2D.points import get_fixed_points

def main(reference, registering):
    # Get fixed points
    points = get_fixed_points(reference)
    # Register
    register(reference, registering, fixed_points=points)


if __name__ == "__main__":
    import napari

    # could grab these from a command line call with sys.argv
    reference = "media/original.jpg"
    registering = "media/rotated.jpg"

    with napari.gui_qt():
        main(reference, registering)
```
then you would remove all of the `gui_qt` calls inside your code.

Another thing you should try to avoid is accessing napari layers _after_ your context managers are closed and the napari viewer is closed:  For instance, in [this line](https://github.com/BrancoLab/imgreg2D/blob/80b5155a66f23142bfb77fb88dcdfa5a3188efce/imgreg2D/points.py#L63), you access `points_layer` on a viewer that has presumably already been closed...  I would instead design that whole [context manager](https://github.com/BrancoLab/imgreg2D/blob/80b5155a66f23142bfb77fb88dcdfa5a3188efce/imgreg2D/points.py#L41-L61) as an independent function that you can call, and have that function return the `points_layer.data` (which seems to be all you're using from the napari layer anyway), rather than the whole layer object itself(which is now orphaned from its viewer).
If you change all those context managers to functions, then you will only need to start up one Qt context per "run", and that should be done at the very highest level, when you start your program (e.g. `test.py`)

**Comment by FedeClaudi** [comment]
Hi,

awesome, thank you so much for taking the time to look into this. I'll see if I can restructure the code to start the Qt context only once, but for the time being the master branch version of napari's `viewer.close()` seems to be working very well. 

Thank you again, 
Federico

**Comment by tlambert03** [comment]
ok, I'm gonna close this issue, since the original request was added in a recent PR, but feel free to leave more comments or open a new issue with any additional thoughts
